CNNVariant Update models, validate scores, cleanup training

lucidtronix opened the pull request:
Code cleanup and technical debt payback, plus updated models.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/CNNScoreVariants.java:
```diff
@@ -372,7 +422,7 @@ private String getVariantDataString(final VariantContext variant) {
     }
 
-    private String getVariantInfoString(final VariantContext variant) {
+    private String getVariantInfoStringOld(final VariantContext variant) {
```
This method appears to be unreferenced now.

lucidtronix replied:
removed

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/CNNScoreVariants.java:
```diff
@@ -92,15 +92,15 @@
 )
 
-public class CNNScoreVariants extends VariantWalker {
+public class CNNScoreVariants extends TwoPassVariantWalker {
```
Nice.

lucidtronix replied:
:)

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/CNNScoreVariants.java:
```diff
@@ -285,6 +279,61 @@ public void apply(final VariantContext variant, final ReadsContext readsContext,
+            scoreScan.useDelimiter("\\n");
+            writeVCFHeader(vcfWriter);
+        } catch (IOException e) {
+            throw new GATKException("Error when trying to write annotated VCF.", e);
```
It could be the score file - might be worth having a nested try/catch for that.

cmnbroad replied:
Also, the Scanner is never closed anywhere.

lucidtronix replied:
it seems createVCFWriter wont throw IOExceptions, so I just fixed the error message. Scanner is now closed.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/CNNScoreVariants.java:
```diff
@@ -285,6 +279,61 @@ public void apply(final VariantContext variant, final ReadsContext readsContext,
+        logger.info("Done scoring variants with CNN.");
+        if ( vcfWriter != null ) {
+            vcfWriter.close();
+        }
```
Should test the Scanner for null and also close that if not.

lucidtronix replied:
done

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/CNNScoreVariants.java:
```diff
@@ -486,5 +495,20 @@ private void setArchitectureAndWeightsFromResources() {
         }
     }
 
+    private void startPythonSession(){
```
Can you rename this to something more specific - maybe one suggestion would be initializePythonArgsAndModel.

cmnbroad replied:
I think it would be a lot simple if all of the code that figures out the values for `weights` and `architecture` was in one place, i.e., all of the if blocks in `setArchitectureAndWeightsFromResources` and `startPythonSession` together in one place. The case where you use Python `None` would require special treatment, but it would be much easier to read if the logic were consolidated.

lucidtronix replied:
renamed and consolidated the fxns.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/CNNScoreVariants.java:
```diff
@@ -248,25 +258,9 @@ public void onTraversalStart() {
-
             pythonExecutor.sendSynchronousCommand(String.format("tempFile = open('%s', 'w+')" + NL, scoreFile.getAbsolutePath()));
             pythonExecutor.sendSynchronousCommand("import vqsr_cnn" + NL);
-
```
Can you move the above `import vqsr_cnn` into the python code like you did with the keras import.

lucidtronix replied:
not easily, because the `executePythonCommand` calls vqsr_cnn.score_and_write_batch so this python session needs the vqsr_cnn module loaded

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/CNNScoreVariants.java:
```diff
@@ -92,15 +92,15 @@
  *   -weights path/to/my_weights.hd5
  * </pre>
  */
+@BetaFeature
```
I've done only superficial review of the python code behind this. Has anyone else ever reviewed it ? We should talk about how to get that done.

lucidtronix replied:
@mbabadi did a review too, but I would be happy for someone to take closer look if there's a volunteer.

lucidtronix commented on scripts/gatkcondaenv.yml.template:
```diff
@@ -43,9 +43,8 @@ dependencies:
   - six==1.11.0
-  - $tensorFlowDependency
-  - tensorflow-tensorboard==0.4.0rc3
+  - tensorflow==1.9.0
```
fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/InfoConcordanceRecord.java:
```diff
@@ -0,0 +1,95 @@
+import org.broadinstitute.hellbender.utils.tsv.TableWriter;
+import org.broadinstitute.hellbender.utils.tsv.TableReader;
+
+public class InfoConcordanceRecord {
```
added

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/InfoConcordanceRecord.java:
```diff
@@ -0,0 +1,95 @@
+            return writer;
+        }
+        catch (IOException e) {
+            throw new UserException(String.format("Encountered an IO exception while reading from %s.", outputTable), e);
```
fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/VcfInfoConcordance.java:
```diff
@@ -0,0 +1,136 @@
+        programGroup=VariantEvaluationProgramGroup.class)
+@DocumentedFeature
+@BetaFeature
+public class VcfInfoConcordance extends AbstractConcordanceWalker {
```
Fixed, and yes I think this is different enough from concordance which evaluates a caller, because this tool is intended for filter scores or annotations.

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/VcfInfoConcordance.java:
```diff
@@ -0,0 +1,136 @@
+    @Argument(doc="A table of summary statistics (true positives, sensitivity, etc.)", fullName="summary", shortName=SUMMARY_SHORT_NAME)
+    protected File summary;
+    @Argument(fullName="eval-info-key", shortName="eval-info-key", doc="Info key from eval vcf", optional=true)
+    protected String evalInfoKey = "CNN_2D";
```
fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/VcfInfoConcordance.java:
```diff
@@ -0,0 +1,136 @@
+    @Argument(fullName="eval-info-key", shortName="eval-info-key", doc="Info key from eval vcf", optional=true)
+    protected String evalInfoKey = "CNN_2D";
+    @Argument(fullName="truth-info-key", shortName="truth-info-key", doc="Info key from truth vcf", optional=true)
+    protected String truthInfoKey = "CNN_2D";
```
fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/VcfInfoConcordance.java:
```diff
@@ -0,0 +1,136 @@
+    public static final String SUMMARY_LONG_NAME = "summary";
+    public static final String SUMMARY_SHORT_NAME = "S";
+    @Argument(doc="A table of summary statistics (true positives, sensitivity, etc.)", fullName="summary", shortName=SUMMARY_SHORT_NAME)
+    protected File summary;
```
fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/InfoConcordanceRecord.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    public static class InfoConcordanceWriter extends TableWriter<InfoConcordanceRecord> {
+        private InfoConcordanceWriter(File output) throws IOException {
```
fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/InfoConcordanceRecord.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    public static class InfoConcordanceReader extends TableReader<InfoConcordanceRecord> {
+        public InfoConcordanceReader(File summary) throws IOException {
```
fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/CNNScoreVariants.java:
```diff
@@ -180,9 +183,11 @@
+    private File scoreFile;
     private String scoreKey;
+    private Scanner scoreScan;
+    VariantContextWriter vcfWriter;
```
yes, fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/CNNScoreVariants.java:
```diff
@@ -228,6 +233,10 @@ protected VariantFilter makeVariantFilter(){
     @Override
     public void onTraversalStart() {
+        if (getHeaderForVariants().getGenotypeSamples().size() > 1) {
+            logger.warn("CNNScoreVariants is a single sample tool, but the input VCF has more than 1 sample.");
```
I think not, there lots of cases where you might want to do this and some evaluation tools add a sample to VCFs for the validated calls, but the VCF is still single sample.

cmnbroad replied:
Ok.

lucidtronix commented:
@cmnbroad thanks for the review back to you!

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+import org.broadinstitute.hellbender.utils.variant.GATKVCFConstants;
+import picard.cmdline.programgroups.VariantEvaluationProgramGroup;
+
+@CommandLineProgramProperties(
```
Doc added, we can use two different keys from 1 VCF or 1 or 2 keys for 2 different vcfs.  Clarified in the doc.

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+    static final String USAGE_SUMMARY = "This tool evaluates info fields from an input VCF against a VCF that has been validated and is considered to represent ground truth.\n";
+    public static final String SUMMARY_LONG_NAME = "summary";
+    public static final String SUMMARY_SHORT_NAME = "S";
+    @Argument(doc="A table of summary statistics (true positives, sensitivity, etc.)", fullName="summary", shortName=SUMMARY_SHORT_NAME)
```
fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+    public static final String SUMMARY_SHORT_NAME = "S";
+    @Argument(doc="A table of summary statistics (true positives, sensitivity, etc.)", fullName="summary", shortName=SUMMARY_SHORT_NAME)
+    protected String summary;
+    @Argument(fullName="eval-info-key", shortName="eval-info-key", doc="Info key from eval vcf", optional=true)
```
fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+    @Argument(doc="A table of summary statistics (true positives, sensitivity, etc.)", fullName="summary", shortName=SUMMARY_SHORT_NAME)
+    protected String summary;
+    @Argument(fullName="eval-info-key", shortName="eval-info-key", doc="Info key from eval vcf", optional=true)
+    protected String evalInfoKey = GATKVCFConstants.CNN_2D_KEY;
```
made them required, I think that is less error-prone.

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+
+    private void infoDifference(VariantContext eval, VariantContext truth) {
+        double evalVal = Double.valueOf((String)eval.getAttribute(this.evalInfoKey));
+        double truthVal = Double.valueOf((String)truth.getAttribute(this.truthInfoKey));
```
fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+    protected String evalInfoKey = GATKVCFConstants.CNN_2D_KEY;
+    @Argument(fullName="truth-info-key", shortName="truth-info-key", doc="Info key from truth vcf", optional=true)
+    protected String truthInfoKey = GATKVCFConstants.CNN_2D_KEY;
+    @Argument(fullName="epsilon", shortName="epsilon", doc="Difference tolerance", optional=true)
```
fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+        return sameRefAllele && containsAltAllele;
+    }
+
+    protected Predicate<VariantContext> makeTruthVariantFilter() {
```
added

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+    @Argument(fullName="epsilon", shortName="epsilon", doc="Difference tolerance", optional=true)
+    protected double epsilon = 0.1;
+    private final EnumMap<ConcordanceState, MutableLong> snpCounts = new EnumMap<>(ConcordanceState.class);
+    private final EnumMap<ConcordanceState, MutableLong> indelCounts = new EnumMap<>(ConcordanceState.class);
```
removed the maps, added eval filters and now explicitly only look at "true positives"

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+            this.logger.warn(String.format("Difference (%f) greater than epsilon (%f) at %s:%d %s:", delta, this.epsilon, eval.getContig(), eval.getStart(), eval.getAlleles().toString()));
+            this.logger.warn(String.format("\t\tTruth info: " + truth.getAttributes().toString()));
+            this.logger.warn(String.format("\t\t Eval info: " + eval.getAttributes().toString()));
+        }
```
Since the intended use here is just for debugging I'd prefer not write them to a file, we could make this a oneTime logger or just remove, although I do think the current behavior could be helpful in the future.

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/InfoConcordanceRecord.java:
```diff
@@ -0,0 +1,101 @@
+    final String evalKey;
+    final String trueKey;
+    final double mean;
+    final double std;
```
fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+    private double indelSumDelta = 0.0;
+    private double indelSumDeltaSquared = 0.0;
+
+    public void onTraversalStart() {
```
fixed

lucidtronix commented on build.gradle:
```diff
@@ -63,8 +63,9 @@ final barclayVersion = System.getProperty('barclay.version','2.1.0')
 final hadoopBamVersion = System.getProperty('hadoopBam.version','7.10.0')
-final tensorflowVersion = System.getProperty('tensorflow.version','1.4.0')
 final genomicsdbVersion = System.getProperty('genomicsdb.version','0.9.2-proto-3.0.0-beta-1+ab5fbe92900259')
+final tensorflowVersion = System.getProperty('tensorflow.version','1.9.0')
```
Yes I have tested and I did not see any issues.

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+        this.logger.info(String.format("SNP average delta %f and standard deviation: %f", snpMean, snpStd));
+        this.logger.info(String.format("INDEL average delta %f and standard deviation: %f", indelMean, indelStd));
+
+        try {
```
fixed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+            this.indelSumDelta += Math.sqrt(deltaSquared);
+            this.indelSumDeltaSquared += deltaSquared;
+        }
+        if (Math.abs(delta) > this.epsilon) {
```
now both VCFs are filtered the same. Should only see INDELs and SNPs here.

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+    protected boolean areVariantsAtSameLocusConcordant(VariantContext truth, VariantContext eval) {
+        boolean sameRefAllele = truth.getReference().equals(eval.getReference());
+        boolean containsAltAllele = eval.getAlternateAlleles().contains(truth.getAlternateAllele(0));
+        return sameRefAllele && containsAltAllele;
```
Since the info fields are site-level values the alt alleles it should be fine to grab the first allele only.

lucidtronix commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordanceIntegrationTest.java:
```diff
@@ -0,0 +1,44 @@
+        final ArgumentsBuilder argsBuilder = new ArgumentsBuilder();
+        argsBuilder.addArgument(AbstractConcordanceWalker.EVAL_VARIANTS_SHORT_NAME, inputVcf)
+                .addArgument(AbstractConcordanceWalker.TRUTH_VARIANTS_LONG_NAME, inputVcf)
+                .addArgument(StandardArgumentDefinitions.REFERENCE_LONG_NAME, b37_reference_20_21)
```
nice catch, removed.

lucidtronix commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordanceIntegrationTest.java:
```diff
@@ -0,0 +1,44 @@
+                .addArgument(EvaluateInfoFieldConcordance.SUMMARY_LONG_NAME, summary.toString());
+        runCommandLine(argsBuilder);
+
+        InfoConcordanceRecord.InfoConcordanceReader reader = new InfoConcordanceRecord.InfoConcordanceReader(summary);
```
fixed

lucidtronix commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordanceIntegrationTest.java:
```diff
@@ -0,0 +1,44 @@
+                .addArgument(StandardArgumentDefinitions.REFERENCE_LONG_NAME, b37_reference_20_21)
+                .addArgument("eval-info-key", GATKVCFConstants.CNN_2D_KEY)
+                .addArgument("truth-info-key", "NOVA_HISEQ_MIX_SMALL")
+                .addArgument("epsilon", "0.01")
```
yes, but now info keys are required, epsilon removed from here.

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+package org.broadinstitute.hellbender.tools.walkers.validation;
+
+import java.io.File;
+import java.nio.file.Path;
```
removed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/InfoConcordanceRecord.java:
```diff
@@ -0,0 +1,101 @@
+
+import htsjdk.variant.variantcontext.VariantContext;
+
+import java.io.File;
```
removed

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,140 @@
+        if (Math.abs(delta) > this.epsilon) {
+            this.logger.warn(String.format("Difference (%f) greater than epsilon (%f) at %s:%d %s:", delta, this.epsilon, eval.getContig(), eval.getStart(), eval.getAlleles().toString()));
+            this.logger.warn(String.format("\t\tTruth info: " + truth.getAttributes().toString()));
+            this.logger.warn(String.format("\t\t Eval info: " + eval.getAttributes().toString()));
```
removed

lucidtronix commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordanceIntegrationTest.java:
```diff
@@ -0,0 +1,44 @@
+        final Path summary = createTempPath("summary", ".txt");
+        final ArgumentsBuilder argsBuilder = new ArgumentsBuilder();
+        argsBuilder.addArgument(AbstractConcordanceWalker.EVAL_VARIANTS_SHORT_NAME, inputVcf)
+                .addArgument(AbstractConcordanceWalker.TRUTH_VARIANTS_LONG_NAME, inputVcf)
```
Added

lucidtronix commented on src/main/python/org/broadinstitute/hellbender/vqsr_cnn/vqsr_cnn/defines.py:
```diff
@@ -5,6 +5,7 @@
 
 TENSOR_MAPS_2D = ['read_tensor']
 TENSOR_MAPS_1D = ['reference']
+# noinspection PyInterpreter
```
accident, I don't know.

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/CNNScoreVariants.java:
```diff
@@ -471,20 +468,33 @@ private String getScoreKeyAndCheckModelAndReadsHarmony() {
+        }
+
+        String getArgsAndModel;
+        if (weights != null && architecture != null) {
```
I agree, but can we save this for the PEP8 PR that is on-deck? It will require several python changes to the python code as well as updates to the WDLs, etc so I would rather only refactor that stuff once.

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/InfoConcordanceRecord.java:
```diff
@@ -0,0 +1,101 @@
+    final double mean;
+    final double std;
+
+    public InfoConcordanceRecord(VariantContext.Type type, String evalKey, String trueKey, double mean, double std) {
```
added

lucidtronix commented:
@cmnbroad thanks for the review.   I think addressed all comments except the arguments/weights simplification, which I would prefer to save for the PEP8 refactor we discussed.  Back to you!

lucidtronix commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordance.java:
```diff
@@ -0,0 +1,179 @@
+        switch (concordanceState) {
+            case TRUE_POSITIVE: {
+                snpCount++;
+                indelCount++;
```
oh man that's bad. Thanks for the catch, fixed.

lucidtronix commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/validation/EvaluateInfoFieldConcordanceIntegrationTest.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Test
+    public void test2Vcfs() throws Exception {
```
Yup that is much nicer. Thanks!

lucidtronix commented:
@cmnbroad back to you. The build.gradle conflict is from the genomicsdb and tensorflow version updates.  I can resolve and rebase if it looks good to you.

cmnbroad commented:
@lucidtronix Looks good now - lets squash and rebase on master and run tests again and should be good to go.
